### DISCUSSION TITLE
Is it possible to trigger eager-loading of transforms for an asset that has already been eager-loaded?

### DISCUSSION THREAD

**Discussion by BigglesZX** [craftcms/cms#10600] [Q&A] [answered]
I'm used to eager-loading Matrix and Asset fields by specifying them as part of an `eagerLoadElements` call in my templates like this:

```twig
{% do craft.app.elements.eagerLoadElements(
    className(entry),
    [entry],
    [
        'myMatrixField',
        'myMatrixField.blockType:assetField',
    ])
%}
```

I'm starting to implement eager-loading for asset transforms, and it would be more maintainable for me to trigger that eager-loading inside the Twig components that render the `picture` and `img` elements, rather than at the page template level, as the components have more specific knowledge of which transforms are needed.

Just as I can trigger eager-loading of entry fields at any time using the `{% do craft.app.elements.eagerLoadElements(...) %}` syntax as above, **is there a way to trigger eager-loading of asset transforms directly from the asset, i.e. when working with an asset object that has itself already been eager-loaded?** At the moment my understanding is that it can only be done at the same time as eager-loading the asset field itself.

I'm thinking something like this:

```twig
{# Warning: fictional! #}
{% do craft.app.elements.eagerLoadAssetTransforms(
    className(entry.myMatrixField[0].assetField[0]),
    [entry.myMatrixField[0].assetField[0]],
    [
        'square',
        { width: 200, height: 100 },
    ])
%}
```

Is that currently possible? Am I missing something obvious? I suppose I could pass an instance of an AssetField into my components instead of an Asset (so I could then call `.withTransforms()` on the field), but it seems preferable not to do that. Any tips would be very much appreciated!

**Comment by BigglesZX** [comment, marked as answer]
I've just found the [`eagerLoadTransforms`](https://docs.craftcms.com/api/v3/craft-services-assettransforms.html#method-eagerloadtransforms) method of the **AssetTransforms** service class, which seems like it might do what I want, but I'm not sure that service is accessible from the template. I'll keep digging...

**Reply by brandonkelly** [reply]
Yep, that’s the one.

```twig
{% do craft.app.assetTransforms.eagerLoadTransforms([asset], [
    'square',
    '...',
]) %}
```

Worth noting that you’re not fully solving N+1 problems when you put off eager-loading until this point though. The whole idea with eager-loading is you do it as early as possible, so you aren’t querying for elements/transforms multiple times.

**Reply by BigglesZX** [reply]
Thanks @brandonkelly! Straight on it as always. Noted your point about N+1 – I’ll have to have a think about the tradeoff of complex template-level eager-loading plan vs excess queries from the component level.

**Reply by BigglesZX** [reply]
Just a note that in Craft 4 the `eagerLoadTransforms` method is now part of the `ImageTransforms` service – the `AssetTransforms` service has been removed.

```twig
{% do craft.app.imageTransforms.eagerLoadTransforms([asset], [
    'square',
    '...',
]) %}
```
